File: 76-minimum-window-substring/76-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ""
        left = 0
        right = 0
        optimalLeft = 0
        optimalRight = 0
        minLen = float("inf")
        counterT = dict()
        for c in t:
            if c in counterT.keys():
                counterT[c] += 1
            else:
                counterT[c] = 1
        requested = len(counterT)
        formed  = 0
        windowsCount = dict()
        while right < len(s):
            currChar = s[right]
            if currChar in windowsCount.keys():
                windowsCount[currChar] += 1
            else:
                windowsCount[currChar] = 1
            if currChar in counterT.keys() and windowsCount[currChar] == counterT[currChar]:
                formed += 1
            while left <= right and formed == requested:
                c = s[left]
                if right - left + 1 < minLen:
                    minLen = right - left + 1
                    optimalLeft = left
                    optimalRight = right
                windowsCount[c] -= 1
                if c in counterT.keys() and windowsCount[c] < counterT[c]:
                    formed -= 1
                left += 1
            right += 1
        if minLen == float("inf"):
            return ""
        else:
            return s[optimalLeft : optimalRight + 1]
        
            

File: 76-minimum-window-substring/76-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";
        HashMap<Character, Integer> countT = new HashMap<>();
        for (Character c : t.toCharArray()) {
            countT.put(c, countT.getOrDefault(c, 0) + 1);
        }
        int uniqueCount = countT.size();
        HashMap<Character, Integer> countS = new HashMap<>();
        int minLeft = 0;
        int minRight = 0;
        int len = s.length() + 1;
        int left = 0;
        int currCount = 0;
        for (int right = 0; right < s.length(); right++) {
            Character c = s.charAt(right);
            countS.put(c, countS.getOrDefault(c, 0) + 1);
            if (countT.containsKey(c) && countT.get(c).equals(countS.get(c))) currCount++;
            while (left <= right && currCount == uniqueCount) {
                if (right - left + 1 < len) {
                    len = right - left + 1;
                    minLeft = left;
                    minRight = right;
                }
                Character leftC = s.charAt(left);
                countS.put(leftC, countS.get(leftC) - 1);
                if (countT.containsKey(leftC) && countS.get(leftC).compareTo(countT.get(leftC)) < 0) {
                    currCount--;
                }
                left++;
            }
        }
        return (len == s.length() + 1) ? "" : s.substring(minLeft, minRight + 1);
    }
}